clang/lib/CIR/CodeGen/CIRGenDecl.cpp

```diff
@@ -322,18 +333,277 @@ void CIRGenFunction::buildVarDecl(const VarDecl &D) {
   // Some function-scope variable does not have static storage but still
   // needs to be emitted like a static variable, e.g. a function-scope
   // variable in constant address space in OpenCL.
-  if (D.getStorageDuration() != SD_Automatic)
-    assert(0 && "not implemented");
+  if (D.getStorageDuration() != SD_Automatic) {
+    // Static sampler variables translated to function calls.
+    if (D.getType()->isSamplerT())
+      return;
+
+    auto Linkage = CGM.getCIRLinkageVarDefinition(&D, /*IsConstant=*/false);
+
+    // FIXME: We need to force the emission/use of a guard variable for
+    // some variables even if we can constant-evaluate them because
+    // we can't guarantee every translation unit will constant-evaluate them.
+
+    return buildStaticVarDecl(D, Linkage);
+  }
 
   if (D.getType().getAddressSpace() == LangAS::opencl_local)
-    assert(0 && "not implemented");
+    llvm_unreachable("OpenCL and address space are NYI");
 
   assert(D.hasLocalStorage());
 
   CIRGenFunction::VarDeclContext varDeclCtx{*this, &D};
   return buildAutoVarDecl(D);
 }
 
+static std::string getStaticDeclName(CIRGenModule &CGM, const VarDecl &D) {
+  if (CGM.getLangOpts().CPlusPlus)
+    return CGM.getMangledName(&D).str();
+
+  // If this isn't C++, we don't need a mangled name, just a pretty one.
+  assert(!D.isExternallyVisible() && "name shouldn't matter");
+  std::string ContextName;
+  const DeclContext *DC = D.getDeclContext();
+  if (auto *CD = dyn_cast<CapturedDecl>(DC))
+    DC = cast<DeclContext>(CD->getNonClosureContext());
+  if (const auto *FD = dyn_cast<FunctionDecl>(DC))
+    ContextName = std::string(CGM.getMangledName(FD));
+  else if (const auto *BD = dyn_cast<BlockDecl>(DC))
+    llvm_unreachable("block decl context for static var is NYI");
+  else if (const auto *OMD = dyn_cast<ObjCMethodDecl>(DC))
+    llvm_unreachable("ObjC decl context for static var is NYI");
+  else
+    llvm_unreachable("Unknown context for static var decl");
+
+  ContextName += "." + D.getNameAsString();
+  return ContextName;
+}
+
+// TODO(cir): LLVM uses a Constant base class. Maybe CIR could leverage an
```

Comment:
I thought about this too, but it's not clear whether that fits the model of CIR, cause we might have constants coming from operations or plain attributes. Have you gave any extra thought about how this could be instead?

Reply:
It's honestly still a bit of a mystery what this interface would look like. Not really sure if it would be beneficial in the end either.

I imagined a wrapper class that would be able to perform type-independent queries regarding constants. For example, we could have some `getConstAttr` method that would return a `#cir.global_view` if the constant was a `cir.global` op, or just a `#cir.int` attribute if the constant was an integer. If this could be an MLIR interface, that will be great, but I don't even know if interfaces can be shared between operations and attributes.